pdns/dnsrecords.cc

```diff
@@ -489,6 +490,195 @@ string EUI64RecordContent::getZoneRepresentation(bool noDot) const
+    if (ret->d_afdlength > 4) {
+      throw MOADNSException("Invalid IP length for IPv4 APL");
+    }
+    bzero(ret->d_ip4, sizeof(ret->d_ip4));
+    for (int i=0; i < 4; i++) {
+      if (i < ret->d_afdlength)
+        pr.xfr8BitInt(ret->d_ip4[i]);
+    }
+  } else if (ret->d_family == APL_FAMILY_IPV6) {
+    if (ret->d_afdlength > 16) {
+      throw MOADNSException("Invalid IP length for IPv6 APL");
+    }
+    bzero(ret->d_ip6, sizeof(ret->d_ip6));
+    for (int i=0; i< 16; i++) {
+      if (i < ret->d_afdlength)
+        pr.xfr8BitInt(ret->d_ip6[i]);
+    }
+  } else
+    throw MOADNSException("Unknown family for APL record");
+
+  return ret;
+}
+
+// Parse backend record
+std::shared_ptr<DNSRecordContent> APLRecordContent::make(const string& zone) {
+  string record;
+  Netmask nm;
+  int bytes;
+  bool done_trimming;
+
+  auto ret=std::make_shared<APLRecordContent>();
+
+  // Strip the optional leading ! (negate)
+  if (zone[0] == '!') {
+    ret->d_n = 1;
+    record = zone.substr(1, zone.length()-1);
+  } else {
+    ret->d_n = 0;
+    record = zone;
+  }
+
+  if (record.find("/") == string::npos) { // Required by RFC section 5
+    throw MOADNSException("Asked to decode '"+zone+"' as an APL record, but missing subnet mask");
+  }
+
+
+  if (record.find("1:", 0) == 0) { // IPv4
+    uint32_t v4ip;
+
+    ret->d_family = APL_FAMILY_IPV4;
+
+    // Ensure that a mask is provided
+
+    // Read IPv4 string into a Netmask object
+    nm = Netmask(record.substr(2, record.length() - 2));
+    ret->d_prefix = nm.getBits();
+
+    if (nm.getNetwork().isIPv4() == 0)
+      throw MOADNSException("Asked to decode '"+zone+"' as an APL v4 record");
+
+    // Section 4.1 of RFC 3123 (don't send trailing "0" bytes)
+    // Copy data; using array of bytes since we might end up truncating them in the packet
+    v4ip = ntohl(nm.getNetwork().sin4.sin_addr.s_addr);
+    bzero(ret->d_ip4, sizeof(ret->d_ip4));
+    bytes  = 4; // Start by assuming we'll send 4 bytes
+    done_trimming = false;
+    for (int i=0; i<4; i++) {
+      ret->d_ip4[3-i] = (v4ip & 255);
+      // Remove trailing "0" bytes from packet and update length
+      if ((v4ip & 255) == 0 and !done_trimming) {
+        bytes--;
+      } else {
+        done_trimming = true;
+      }
+      v4ip = v4ip >> 8;
+    }
+    ret->d_afdlength = bytes;
+
+  } else if (record.find("2:", 0) == 0) { // IPv6
+    ret->d_family = APL_FAMILY_IPV6;
+
+    // Parse IPv6 string into a Netmask object
+    nm = Netmask(record.substr(2, record.length() - 2));
+    ret->d_prefix = nm.getBits();
+
+    if (nm.getNetwork().isIPv6() == 0)
+      throw MOADNSException("Asked to decode '"+zone+"' as an APL v6 record");
+
+    // Section 4.2 of RFC 3123 (don't send trailing "0" bytes)
+    // Remove trailing "0" bytes from packet and reduce length
+    bzero(ret->d_ip6, sizeof(ret->d_ip6));
+    bytes = 16; // Start by assuming we'll send 16 bytes
+    done_trimming = false;
+    for (int i=15; i>=0; i--) {
```

Comment:
it would be probably more sensible to write this as `int i=0; i<16; i++` and `ret->d_ip6[15-i]`;